README.txt
SnowMicroPyn Version 0.0.20 alpha
=================================

About:
------
SnowMicroPyn is a Python based reader for SnowMicroPen binary files. The Software
is open source and licensed under GNU GPL terms of conditions.
Feel free to use, modify and/or extend the code.

The program is still an alpha revision. Please report bugs, see section "Contact".
Development takes place mainly on and for Linux systems, thus bugs might occur in
in MacOSX and Win versions.

Newest software releases can be found here:
http://sourceforge.net/projects/pyntreader

Features:
---------
-Read and plot .pnt Files
-Export measurement, infos and plots into human readable
-View and save measurement GPS locations using Google Static API
-Noise, drift and offset analysis
-batch mode
-super pose and subtract plots
-frequency analysis

Requirements:
-------------
A) Using a Python Interpreter:
	-tested on python 2.7.x
	-use wxPython 3.0.0
	-use matplotlib == 1.2.2 (version >= 1.2.2 has a bug in zoom function)
	-use scipy == 13.0
	-see ./src/extensions/requirements.txt
	-use numpy >= 1.8.1
	
B) Stand alone binaries:
	- locate binary (.bin/.app/.exe) in the same folder as artwork directory
	- no other dependencies are known

Installation:
-------------
A) Using a Python Interpreter:
	-install Python 2.7.x (python.org)
	-it is recommended to use virtualenv, see http://blog.dbrgn.ch/2012/9/18/virtualenv-quickstart/
	-install dependencies using pip
		pip install -r requirements.txt
		wxpython >= 2.8 can be installed from wxpython.org
	-run "python SnowMicroPyn.py" in terminal
	-or in Ubuntu make file executable: "chmod +x PyNTReader.py" and run ./PyNTReader from terminal, respectively double click the file 

B) Stand alone binaries:

	I) Linux
		compiled and tested on Ubuntu 14.04 64bit
		-make file executable: chmod +x PyNTReader
		-double click executable or type /path/to/file/PyNTReader
		-optionally pass file names as command line argument to open them directly, or try "open with"
	II) MacOS
		compiled and tested on MacOSX 10.9.5 64bit
		-double click .app or executer PyNTReader in terminal
	III) Windows
		compiled and tested on Win7 64bit and WinXP 32bit
		-double click .exe, or press on "open with" on a .pnt file 

Contact:
--------
For questions and bug reports please contact sascha.grimm@slf.ch

Developer:
-----------
Sascha Grimm
WSL Institute for Snow and Avalanche Research SLF
Team Snow Physics
Flüelastrasse 11
CH-7260 Davos


CHANGELOG.txt
--------------------------
SnowMicroPyn 0.1.1
--------------------------

ToDo:
-manual layer insertion -proksch
-2d log(median) profile, frame 2.5 mm overlap 50% -proksch
-option for plot rotation 90degs
-force drops statistics -denes
-shot noise parameters check

2016/07/24
- implemented log file creation /path/to/src/.SnowMicroPyn.log
- removed unused/unclean functionalities (Henning)

2014/08/18
- implemented matlpotlib.rcParams["agg.path.chunksize"] = 10000 to improve plotting speed and prevent buffer overflow when plotting large data sets
- added new autozoom to graphoptions
- save image as pdf instead of png

2014/08/15
- version 0.0.20 alpha
- show legend beside plot instead of in plot

2014/08/07
- improved drift/noise/mean tool: boarders dragable now

2014/07/18
- implemented save data in super position viewer

2014/07/17
- implemented check if file is already opened (no duplicated files)

2014/07/15
- implemented exception in getSurface for Butterworth filter (was not working in scipy 0.14)

2014/07/10
- version 0.0.19 alpha
- allowed negative and comma values for axes in graph options
- new variable self.pathOpen as default file opening path (location of last opened file)
- new variable self.pathSave as default save path

2014/04/29
- gps coordinates sign bug fixed
- fixed bug, where noise, drift offset could not be shown

2014/04/04
- implemented options for subtracted median in options window
- implemented shot noise and added shot noise save options

2014/04/02
-fixed surface tool visibility bug
-introduced overwrite prompt to some save functions
-arrow keys can be used to switch plot
-fixed indexing bug in shotnoise()

2014/04/01
-improved mean.py
-introduced view menu, exported functions from data menu

2014/03/28
-use scroll wheel to zoom in/out of plot
-up/down/left/right to navigate through open files
-ctrl+ left click moves surface to location
-shift + left click moves ground to location

2014/03/27
-implemented average menu to data (/extensions/mean.py)
-started to use click event on canvas
-left click on plot shows x//y coordinates

2014/03/24
-implemented average curves to superposition viewer
-implemented log for y axis to graph options, default = True
-implemented subtract median to view menu

2014/03/21
-call class graphoptions with .show() -> parent.draw_figure actualizes parent plot

2014/03/20
-version 0.0.14 alpha
-add ground level analog to surface
-add overload to info screen

2014/03/17
-version 0.0.13 alpha
-add option to hide legend in super position viewer

2014/03/16
-implemented graph options to super position viewer
-hide surface tool if not checked
-get_surface returns now max instead of a rounded value, if no surface was found
-created a pyinstaller build script to create build archives

2014/02/24
-bug fix in OnClose()

2014/02/21
-renamed software to SnowMicroPyn

2014/02/19
-fixed bug in export surface
-fixed surface tool bug

2014/02/18
-Additions in Super Position Viewer:
	-subtract plot has same color as corresponding curve
	-legend for subtracted graphs
	-rsme is shown mathematics.rsme(x_ref,x_sub, norm)

2014/02/17
-version 0.0.11 alpha
-fixed surface tool inactivation bug for windows
-maximize super position viewer at start
-deactivated surface tab in graph options
-cancel possibility in GPS viewer if no coordinates available
-precision option in save options for ascii data
-enabled keyboard short cuts for windows
-correct exec_path detection (for exe)

2014/02/16
-version 0.0.10 alpha
-super position viewer: show only basename in reference selection
-fixed bug in file number text control in tool bar
-use Ubuntu icons for tool bar in all operating systems

2014/02/15
-changed standard gradient down sampling factor to 1000

2014/02/10
-fixed logo location bug

2014/02/10
-changed version to 0.0.9 alpha
-changed getsurface algorithm

2014/02/08
-improved filter function

2014/02/08
-implemented experimental butterworth low pass filter and automatic
 cut off frequency search using residual analysis

2014/02/07
-upgrade to 0.0.8 alpha
-implemented open files as command line arguments
 -> "open with"
-implemented error caching to open file function
-reneamed pnt header dict keys with units
	-> header infos contain units
-cleaned code in smp.py
-introduced savezoom() to updatefigure()
-disable preferences when no file is loaded

2014/02/06
-upgraded version to 0.0.7 alpha
-implemented auto zoom boolean to draw_plot
	-> zoom ratio is kept when changing preferences
-implemented gradient to analysis menu
-better surface detection algorithm
-implemented new surface detection algorithm based on 2nd deviation
-fixed bug: smooth in Super Position Viewer works now
-removed subplot from navigation tool bar
-added additional tabs and options to graph options
-implemeted manual surface correction to tool bar
-outsourced getsurface to mathematics.py
-outsourced linfit to mathematics.py
-implemented "keep zoom"
-added legend to super position viewer

2014/02/04
-improved getData from smp.py
-replaced smooth with downsample in getsurface
-introduced arg boolean "show" to draw_plot -> circumvents double plotting while saving plot
-implemented "subtract plot" to SuperPosition
-improved surface detection

2014/02/03
-update version to 0.0.6 alpha
-new class SuperPosition in menus.py
-introduced SuperPosition (not fully developed yet) to data menu

2014/02/02
-added more colors and styles to plot options
-moved class checklstctr from map to menus

2014/01/31
-changed version number to 0.0.5 alpha
-introduced new graph options windows with enhanced functionality.

2014/01/30
-changed "preferences" to "graph options" and moved to data menu
-renamed "view" to "analysis"
-introduced save options to single file save method
-changed graph options tool bar icon
-changed quit icon to cross mark (probably nicer in windows)
-renamed "Map" to "GPS Map View"
-removed "save" from mpl tool bar
-introduced save all and save single to tool bar

src/extensions/map.py
import wx
from cStringIO import StringIO
import os, string, urllib2
from menus import CheckListCtrl

#globals
map_types = ["Satellite","Road Map","Hybrid","Terrain"]
slf_coords = [46.812151, 9.847202] 
icon = "http://tinyurl.com/p93j947"

class Map(wx.Frame):
    """main Frame"""
    def __init__(self, parent, id, files=[], title="Measurement Locations"):
        wx.Frame.__init__(self, parent, id, title, size = (800, 500))
        
        self.picture = wx.StaticBitmap(self)
        
        self.files = files
        self.center = 0
        self.label = []
        self.zoom = 14
        self.maptype = "satellite"
        
        
        if self.internet_on():
            self.createItems()
            if  self.isListChecked():
                self.bmp = self.getImage()
                self.updatePicture()
                self.createLayout()
                self.Centre()
                self.Show()
        else:
            self.Destroy() 
        
    def getImage(self, resolution="250x250"):
        """ capture image from google static api"""
        
        self.GetCenter()
        base = "http://maps.googleapis.com/maps/api/staticmap?"
        center = "center=" + str(self.center[0]) + "," + str(self.center[1])
        size = "&size=" + resolution
        zoom = "&zoom=" + str(self.zoom)
        maptype = "&maptype=" + self.maptype
        scale = "&scale=2"
        marker = self.GetLabel()
        sensor = "&sensor=false"
        url = base + center + size + zoom + scale + maptype + marker + sensor
        imagedata = StringIO(urllib2.urlopen(url).read())
        self.stream = wx.ImageFromStream(imagedata)
        bmp = wx.BitmapFromImage(self.stream)
        return bmp
    
    def updatePicture(self):
        """update image"""
        self.picture.SetFocus()
        self.picture.SetBitmap(self.bmp)
        
    def createItems(self):
        """create all items"""
        #menu
        self.menubar = wx.MenuBar()  
        self.fileMenu = wx.Menu()
        qms = self.fileMenu.Append(wx.ID_SAVE, '&Save Image')
        self.Bind(wx.EVT_MENU, self.OnSave, qms)
        qec = self.fileMenu.Append(wx.ID_ANY, '&Export Coordinates')
        self.Bind(wx.EVT_MENU, self.OnExport, qec)
        qme = self.fileMenu.Append(wx.ID_EXIT, '&Quit')
        self.Bind(wx.EVT_MENU, self.OnQuit, qme)
        self.menubar.Append(self.fileMenu, '&File')
        self.SetMenuBar(self.menubar) 
        
        #image options
        self.zoomTxt = wx.StaticText(self, 200, "Zoom Factor")
        self.slider = wx.Slider(self, 300, value=14, minValue=1, maxValue=20, style=wx.SL_HORIZONTAL,size=(500,-1))
        self.combo = wx.ComboBox(self, 500, "Satellite",choices=map_types, style=wx.CB_READONLY,size=(180,-1))
        
        wx.EVT_SLIDER(self,300,self.OnEvent)
        wx.EVT_COMBOBOX(self,500,self.OnEvent)
        
        
        #list
        self.list = CheckListCtrl(self,size=(675,500))
        self.list.InsertColumn(0, 'Label')
        self.list.InsertColumn(1, 'File Name')
        self.list.InsertColumn(2, 'Latitude')
        self.list.InsertColumn(3, 'Longitude')
        self.list.InsertColumn(4, 'PDOP')
        self.list.SetColumnWidth(0, 50)
        self.list.SetColumnWidth(1, 150)
        self.list.SetColumnWidth(2, 200)
        self.list.SetColumnWidth(3, 200)
        self.list.SetColumnWidth(4, 50)
        
        self.list.OnCheckItem = self.OnCheckItem
        
        wx.EVT_CHECKBOX(self.list,600,self.OnEvent)
        
        #create list entries  
        letters = string.ascii_uppercase
        i = 0
        num_labels = 0       
        for entry in self.files:
            lat = self.files[i].header["Latitude"]
            long = self.files[i].header["Longitude"]
            name = os.path.basename(self.files[i].filename)
            pdop = self.files[i].header["PDOP"]
              
            if -180<=lat<=180:
                try:
                    label = letters[num_labels]
                    num_labels += 1
                except:
                    label = letters[0]
                    num_labels = 1
                
            else:
                label = "-"
            
            self.label.append(label)
            self.list.InsertStringItem(i,label)
            if not label == "-": self.list.CheckItem(i, True)    
            self.list.SetStringItem(i,1,name)
            self.list.SetStringItem(i,2,str(lat))
            self.list.SetStringItem(i,3,str(long))
            self.list.SetStringItem(i,4,"%.1f"%pdop)
            i+=1
    
    def isListChecked(self):
        if [index for index in range(self.list.ItemCount)
                if self.list.IsChecked(index)] == []:
                    return self.nothingToShow()
        else:
            return True
        
    def OnCheckItem(self, index, flag):
        """checking item event"""
        if self.files[index].header["Latitude"] == 0.0 or self.files[index].header["Latitude"] > 180:
            self.list.CheckItem(index, False)
        else:
            self.zoom = self.slider.GetValue()
            map = self.combo.GetValue()
            "Satellite","Road Map","Hybrid","Terrain"
            if map == "Satellite":  self.maptype = "satellite"
            if map == "Road Map":   self.maptype = "roadmap"
            if map == "Hybrid":     self.maptype = "hybrid"
            if map == "Terrain":    self.maptype = "terrain"
            
            
            self.bmp = self.getImage()
            self.updatePicture()
            self.createLayout()
            self.Update()
           
    def GetCenter(self): 
        """get center of active coordinates"""           
        try:
            sum_lat = 0
            sum_long = 0
            i = 0
            for entry in range(len(self.files)):
                if self.list.IsChecked(entry):
                    sum_lat += self.files[entry].header["Latitude"]
                    sum_long += self.files[entry].header["Longitude"]
                    i += 1
            self.center = [sum_lat/float(i),sum_long/float(i)]
                    
        except:
            self.center = slf_coords
            pass
     
    def GetLabel(self):
        """create labels for google static api"""
        try:
            label = []
            for entry in range(len(self.files)):
                if self.list.IsChecked(entry):
                    label.append("&markers=size:mid|color:red%7Clabel:" + self.label[entry] + "%7C" + str(self.files[entry].header["Latitude"]) + "," + str(self.files[entry].header["Longitude"]))                    
        except:
            label = "&markers=size:mid|color:red%7Clabel:" + "A" + "%7C" + str(slf_coords[0]) + "," + str(slf_coords[1])
            pass
        if label == []:
            label = "&markers=icon:"+ icon + "|" + str(slf_coords[0]) + "," + str(slf_coords[1])
                
        marker = "".join(label)
        
        return marker
            
    def createLayout(self):
        """set window layout"""
        
        sizer = wx.BoxSizer()
        
        leftbox = wx.BoxSizer(wx.VERTICAL)   
        leftbox.Add(item=self.picture, flag=wx.ALL, border=10)
        leftbox.Add(item=self.zoomTxt, flag=wx.CENTER,border=10)
        leftbox.Add(item=self.slider, flag=wx.LEFT,border=10)       
        leftbox.Add(item=self.combo, flag=wx.CENTER,border=10)
        leftbox.AddSpacer(10)
        
        rightbox = wx.BoxSizer(wx.VERTICAL)
        rightbox.Add(self.list,flag=wx.ALL, border=10)
        
        sizer.Add(leftbox)
        sizer.Add(rightbox)
        self.SetSizerAndFit(sizer)

    def internet_on(self):
        """check if internet connection can be established"""
        try:
            urllib2.urlopen('http://74.125.228.100',timeout=2)
            return True
        except urllib2.URLError:
            message = """No Internet Connection established.
                        Can't show google maps."""
            caption = "Warning"
            warning = wx.MessageDialog(self,message,caption,wx.OK) 
            warning.ShowModal()
            warning.Destroy()
            return False
    
    def OnEvent(self,e):
        """define user events"""
        self.zoom = self.slider.GetValue()
        map = self.combo.GetValue()
        "Satellite","Road Map","Hybrid","Terrain"
        if map == "Satellite":  self.maptype = "satellite"
        if map == "Road Map":   self.maptype = "roadmap"
        if map == "Hybrid":     self.maptype = "hybrid"
        if map == "Terrain":    self.maptype = "terrain"
        
        self.bmp = self.getImage()
        self.updatePicture()
        self.createLayout()
        self.Update()

    def OnSave(self,e):
        """save button functionality"""
        
        file_choices = "PNG (*.png)|*.png"
        
        dlg = wx.FileDialog(
            self, 
            message="Save Map as...",
            defaultDir=os.getcwd(),
            defaultFile="Map.png",
            wildcard=file_choices,
            style=wx.SAVE|wx.OVERWRITE_PROMPT|wx.FD_CHANGE_DIR)
        
        if dlg.ShowModal() == wx.ID_OK:
            path = dlg.GetPath()
            image = self.getImage("800x800")
            image.SaveFile(path, wx.BITMAP_TYPE_PNG)
            
        dlg.Destroy()

    def OnQuit(self,e):
        print "exit"
        self.Destroy()
        
    def OnExport(self,e):
        """export coordinates"""
        
        file_choices = "COORD (*.coord)|*.COORD"
        
        dlg = wx.FileDialog(
            self, 
            message="Save Coordinates as...",
            defaultDir=os.getcwd(),
            defaultFile="SMP_Coordinates.txt",
            wildcard=file_choices,
            style=wx.SAVE|wx.OVERWRITE_PROMPT)
        
        if dlg.ShowModal() == wx.ID_OK:
            path = dlg.GetPath()
            f = open(path, 'w')
            f.write('#Automatic written Coordinates by PyNTReader\n'
                '#SLF Institute for Snow and Avalanche Research\n\n'
                "#Filename\tLongitude\tLatitude\tPdop\n")

            for i in range(len(self.files)):
                line = "%s\t%f\t%f\t%.2f\n" %(os.path.basename(self.files[i].filename),self.files[i].header["Longitude"],self.files[i].header["Latitude"],self.files[i].header["PDOP"])
                f.write(line)
            f.close()
            
        dlg.Destroy()
        
    def nothingToShow(self):
            caption = "Warning"
            message = """Oooops, no coordinates to display.\nShow SLF in Davos, instead?"""
            warning = wx.MessageDialog(None,
                                       message,
                                       caption,
                                       wx.YES_NO|wx.ICON_EXCLAMATION)
             
            return warning.ShowModal() == wx.ID_YES
            
        
class MyApp(wx.App):
    def OnInit(self):
        frame = Map(None, -1)
        self.SetTopWindow(frame)
        return True

if __name__=="__main__":
    app = MyApp(0)
    app.MainLoop()

src/extensions/mean.py
import matplotlib.pyplot as plt
import numpy as np

class Drift(object):
    def __init__(self,x,y):
        self.dataX = x
        self.dataY = y   
        
        self.fig, (self.ax, self.ax2) = plt.subplots(2, 1)
        self.fig.canvas.set_window_title('SMP Mean Value, Drift and Noise Analysis Tool') 
        
        self.fig.subplots_adjust(hspace=0.4 )
        
        self.clicks = 0
        self.mouseState = False
        
        self.v1  = self.ax.axvline(x = x[0], color = "red", visible = False, linewidth=3, alpha=0.6)
        self.v2 = self.ax.axvline(x = x[-1], color = "red", visible = False, linewidth=3, alpha=0.6)

        self.ax.set_title('Original Data')
        self.ax.set_xlabel("Depth [mm]")
        self.ax.set_ylabel("Force [N]")
        self.ax2.set_title("Selected Range")
        self.ax2.set_ylabel("Force [N]")
        self.ax2.set_xlabel("Depth [mm]")
        self.line, = self.ax.plot(self.dataX, self.dataY, picker=5)  # 5 points tolerance
        
        self.fig.canvas.mpl_connect('pick_event', self.onPick)
        self.fig.canvas.mpl_connect('button_press_event', self.mouseDown)
        self.fig.canvas.mpl_connect('motion_notify_event', self.mouseMotion)
        self.fig.canvas.mpl_connect('button_release_event', self.mouseUp)
        
        plt.show()
     
    def mouseDown(self, event):
        if self.v1 in self.fig.hitlist(event) or self.v2 in self.fig.hitlist(event):
            self.mouseState = True
    
    def mouseMotion(self, event):
        if self.mouseState:
            x, y = event.xdata, event.ydata
            if x is None or y is None:
                self.picked = None
                return
            
            if self.v1 in self.fig.hitlist(event):
                self.picked = "v1"
            
            if self.v2 in self.fig.hitlist(event):
                self.picked = "v2"
            
            self.update(event)
            
    def mouseUp(self, event):
        self.mouseState = False
        self.picked = None
        
    def onPick(self,event):
        self.clicks += 1

        x = event.mouseevent.xdata
        y = event.mouseevent.ydata
        
        distances = np.hypot(x-self.dataX[event.ind], y-self.dataY[event.ind])
        indmin = distances.argmin()
        self.index = event.ind[indmin]
        self.update(event)
        
    def update(self, event):

        if self.index is None: return
    
        if self.clicks == 1:
            v = ([self.dataX[self.index],self.dataX[self.index]],[0,10])
            self.v1.set_data(v)
            self.v1.set_visible(True)
            
        if self.clicks == 2:
            self.clicks += 1
            v = ([self.dataX[self.index],self.dataX[self.index]],[0,10]) 
            self.v2.set_data(v) 
            self.v2.set_visible(True)
            
        if self.clicks >= 2:
            
            self.ax2.clear()
            self.ax2.set_title("Selected Range")
            self.ax2.set_ylabel("Force [N]")
            self.ax2.set_xlabel("Depth [mm]")
            
            if self.mouseState:
                x = event.xdata
                v = ([x,x],[0,10])
                if self.picked == "v1":
                    self.v1.set_data(v)
                if self.picked == "v2":
                    self.v2.set_data(v)
                
            args = [self.v1.get_data()[0][0],self.v2.get_data()[0][0]] 
            
            min = np.where(self.dataX >= np.amin(args))[0][0]
            max = np.where(self.dataX >= np.amax(args))[0][0]
            
            x = self.dataX[min:max]
            y = self.dataY[min:max]
            
            self.ax2.plot(x,y)
            self.ax2.set_xlim((np.amin(x),np.amax(x)))
            self.ax2.set_title("Selected Range")
            self.ax2.set_ylim((np.amin(y),2*np.amax(y)))
            
            self.linFit(x,y)
            
        self.fig.canvas.draw()
        
    def linFit(self,x,y):
        mean = np.mean(y)
        dev = np.std(y)
        m,c = np.polyfit(x, y, 1)
        y_fit = x*m+c
        std = np.sqrt(np.mean((y-y_fit)**2))
        self.ax2.plot(x,y_fit, "r--",x,y_fit+std, "b:",x,y_fit-std,"b:")
        self.ax2.text(0.05, 0.9, 'Mean: (%.2e +- %.2e) N\nSlope: %.2e N/m\nStd: %.2e N' %(mean, dev, m*1000, std), transform=self.ax2.transAxes, va='top')        

if __name__ == "__main__":
    test = Drift(np.arange(0,10000), np.random.normal(0,1,10000))

src/pnt.py
"""
pnt.py contains the Pnt object to parse, manipulate, and save
binary SnowMicroPen measurement .pnt files.

Contact: sascha.grimm@dectris.com, loewe@slf.ch

example:

import pnt
import matplotlib.pylab as plt

p = pnt.Pnt("/path/to/file.pnt", verbose=True) # open a .pnt file, set verbosity

print p.header # print meta data dict
p.writeHeader("header.txt") # write header data as txt file

plt.plot(p.data[:,0], p.data[:,1]); plt.show() # plot force displacement curve

p.fromFile("other/filename.pnt") # load new file into object p

p.header["Comment"] = "custom comment" # modify meta data "Comment"
p.writePnt() # write modified meta data to .pnt file
"""

import struct
import os
import numpy

__author__ = "SasG"
__date__ = "16/03/28"
__version__ = "0.1.0"

#.pnt structure parameter table containing:
# [parameter name, binary type, start bit, bit length, physical unit]
PARAMTABLE = [["Version","H",0,2, None], # .pnt file version
			["Tot Samples","i",2,4, None], # number of samples ??? maybe num of force samples + number of t samples ???
			["Samples Dist [mm]","f",6,4, "mm"], # displacement betweene consecutive force samples
			["CNV Force [N/mV]","f",10,4, "N/mV"], # force conversion factor
			["CNV Pressure [N/bar]","f",14,4, "N/bar"], # pressure conversion factor
			["Offset [N]","H",18,2, "N"], # force offset
			["Year","H",20,2, "y"], # acquitision year
			["Month","H",22,2, "m"], # acquisition month
			["Day","H",24,2, "d"], # acquisition day
			["Hour","H",26,2, "h"], # acquitiion hour
			["Min","H",28,2, "min"], # acquisition minute
			["Sec","H",30,2, "s"], # acquisition seconds
			["X Coord","d",32,8, "deg"], # ???
			["Y Coord","d",40,8, "deg"], # ???
			["Z Coord","d",48,8, "deg"], # ???
			["Battery [V]","d",56,8, "V"], # internal battery voltage
			["Speed [mm/s]","f",64,4, "mm/s"], # probing velocity
			["Loopsize","l",68,4, None], # ???
			["Waypoints","10l",72,40, None], # ???
			["Calstart","10H",112,20, None], # ???
			["Calend","10H",132,20, None], # ???
			["Length Comment","H",152,2, None], # length of comment chars
			["Comment","102s",154,102, None], # comment string
			["File Name","8s",256,8, None], # original file name
			["Latitude","f",264,4, "deg"], # gps latitude
			["Longitude","f",268,4, "deg"], # gps longitude
			["Altitude [cm]","f",272,4, "cm"], # gps altitude
			["PDOP","f",276,4, None], # ???
			["Northing","c",280,1, None], # northing string
			["Easting","c",281,1, None], # easting string
			["Num Sats","H",282,2, None], # number of gps satelites
			["Fix Mode","H",284,2, None], # gps fix mode
			["GPS State","c",286,1, None], # gps state
			["reserved 1","x",187,1, None], # reserved
			["X local","H",288,2, "deg"], # ???
			["Y local","H",290,2, "deg"], # ???
			["Z local","H",292,2, "m"], # ???
			["Theta local","H",294,2, "deg"], # ???
			["reserved 2","62x",296,62, None], # reserved
			["Force Samples","l",358,4, None], # numbers of force samples
			["Temperature Samples","l",362,4, None], # number of temperature samples
			["Kistler Range [pC]","H",366,2, "pC"], # Kistler range
			["Amp Range [pC]","H",368,2, "pC"], # amplifier range
			["Sensitivity [pC/N]","H",370,2, "pC/N"], # amplifier sensitivity
			["Temp Offset [K]","h",372,2, "Celsius"], # temperature offset
			["Hand Op","H",374,2, None], # ???
			["Diameter [um]","l",376,4, "um"], # probing tip diameter
			["Overload [N]","H",380,2, "N"], # force overload value
			["Sensor Type","c",382,1, None], # sensor type
			["Amp Type","c",383,1, None], # amp type
			["SMP Serial","H",384,2, None], # SMP serial
			["Length [mm]","H",386,2, "mm"], # SMP rod length
			["reserved 3","4x",388,4, None], # reserved
			["Sensor Serial","20s",392,20, None], # force sensor serial
			["Amp Serial","20s",412,20, None], # amp serial
			["reserved 4 ","80x",432,80, None]] # reserved

class Pnt():
	def __init__(self, filename, verbose=False):
		"""
		Create Pnt object from given filename.
		Input:
			-filename: path to file to parse
		Returns:
			-self.filename: Path to .pnt file [str]
			-self.data: force and displacement data [numpy array]
			-self.header:
			-self.units:
		Private:
			-self.__verbose__: print some additional information
		"""
		self.__verbose__ = verbose # verbosity
		self.units = [param[4] for param in PARAMTABLE]

		# create Pnt object from file
		self.fromFile(filename)


	def fromFile(self, fname):
		"""
		create Pnt object from file and return self
		"""
		self.filename = fname # path to .pnt file
		raw = self.getRaw() # read raw data
		self.header = self.getHeader(raw) # read header as dict
		self.data = self.getData(raw) # read data points
		return self

	def __str__(self):
		return self.printHeader() + "\n" + self.printData()

	def setVerbose(self, verbose=True):
		"""
		set verbosity
		"""
		self.__verbose__ = verbose
		return self.__verbose__

	def getRaw(self):
		"""
		Return raw data from .pnt file
		"""
		with open(self.filename,"rb") as f:
			raw = f.read()
			if self.__verbose__:
				print "Read %d bytes in %s" %(len(raw), self.filename)
			f.close()

		return raw

	def getHeader(self, raw=None):
		"""
		Read Header from raw data using the parameter table
		return header dict {param:value}, and list of paramater units
		"""
		if not raw:
			raw = self.getRaw()
		#read header values and return dict vwith parameter name : value
		header = {}
		for (key, fmt, start, length, unit) in PARAMTABLE:
			fmt = ">" + fmt
			end = start + length
			try:
				value = struct.unpack(fmt,raw[start:end])
				if len(value) == 1: # single value
					header[key] = value[0]
				elif len(value) == 0: # empty value
					header[key] = None
				else:
				 	header[key] = value
			except:
				raise IOError("Could not decode header entry %s at byte %d" %(key,start))

		#format some strings
		if header["Length Comment"]:
			header["Comment"] = header["Comment"][:header["Length Comment"]]
		else:
			header["Comment"] = ""

		header["File Name"] = header["File Name"].split("\x00")[0]
		header["Amp Serial"] = header["Amp Serial"].split("\x00")[0]
		header["Sensor Serial"] = header["Sensor Serial"].split("\x00")[0]
		header["Northing"] = header["Northing"].split("\x00")[0]
		header["Easting"] = header["Easting"].split("\x00")[0]
		header["GPS State"] = header["GPS State"].split("\x00")[0]

		# add sign to coordinates if required
		if header["Northing"] == "S":
			header["Latitude"] = - header["Latitude"]
		if header["Easting"] == "W":
			header["Longitude"] = - header["Longitude"]

		if self.__verbose__:
			print "decoded header of %s" %self.filename

		return header

	def printHeader(self):
		"""
		return header entries in alphanumerical order as formatted string
		"""
		msg = "Header:"
		for key,value in sorted(self.header.items()):
			msg += "\n{0}: {1}".format(key.ljust(15) , value)
		return msg

	def printData(self):
		"""
		return force displacement data formatted string
		"""
		msg = "x [mm]\ty [N]"
		for (x,y) in zip(self.data[:,0],self.data[:,1]):
			msg += "\n{:.3f}\t{:.3f}".format(x,y)
		return msg

	def writeHeader(self, fname=None):
		""""
		write header data to text file fname
		return fname
		"""
		if not fname:
			fname = self.filename.replace(".pnt", ".txt")
		fname = os.path.join(os.getcwd(),fname)

		with open(fname, "w") as f:
			f.write(	"#Automatic written Header from SnowMicroPen .pnt binary\n"
						"#SLF Institute for Snow and Avalanche Research\n#\n"
						"##################################\n#Header:\n##################################\n")

			for entry, value in sorted(self.header.items()):
				line = "%s %s\n" %(entry.ljust(15), str(value))
				f.write(line)
			if self.__verbose__:
				print "wrote header of %s to %s" %(self.filename, fname)
			f.close()

		return fname

	def getData(self, raw=None):
		"""
		get force and displacement data from pnt raw data
		return numpy ndarray:
		displacement [mm]: axis 0
		force [N]: axis 1
		"""
		if not raw:
			raw = self.getRaw()

		try:
			start = 512
			end = self.header["Force Samples"] * 2 + start
			frmt = ">" + str(self.header["Force Samples"]) + "h"
			data = struct.unpack(frmt, raw[start:end])
		except:
			raise IOError("Error while reading data points in %s" %self.filename)

		dx = self.header["Samples Dist [mm]"]
		data_x = numpy.arange(0, len(data)) * dx
		data_y = numpy.asarray(data) * self.header["CNV Force [N/mV]"] # convert mV to Force
		data = numpy.column_stack([data_x,data_y])

		if self.__verbose__:
			print "Read %d data points in %s" %(len(data_y),self.filename)
		return data

	def writePnt(self, fname=None):
		"""
		write .pnt file from Pnt class header infos and data
		return fname
		"""
		if not fname:
			fname = self.filename
		fname = os.path.join(os.getcwd(),fname)

		buff = "" # buffer string

		data = self.data[:,1] / self.header["CNV Force [N/mV]"] # convert N to mV

		# adapt header entries
		self.header["Force Samples"] = len(data) # correct the number of force samples if data have been modified
		self.header["Length Comment"] = len(self.header["Comment"])

		#pack header to buffer
		for (key, fmt, start, length, unit) in PARAMTABLE:
			fmt = ">" + fmt # big endian
			if type(self.header[key]) is tuple:
				buff += struct.pack(fmt, *self.header[key])
			elif "x" in fmt: # reserved spaces
				buff += length * struct.pack(">"+"s","\x00")
			else:
				buff += struct.pack(fmt, self.header[key])

		#pack data array to buffer
		buff += struct.pack(">"+len(data)*"h", *data)

		#write buffer to file
		with open(fname,"w+b") as f:
			f.write(buff)
			f.close()

		if self.__verbose__:
			print "wrote %d bytes to %s" %(len(buff),fname)

		return fname
